File: app/src/main/java/com/sjn/taggingplayer/db/TotalSongHistory.java
package com.sjn.taggingplayer.db;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

@Accessors(prefix = "m")
@AllArgsConstructor(suppressConstructorProperties = true)
@NoArgsConstructor
@Getter
@Setter
public class TotalSongHistory extends RealmObject {
    @PrimaryKey
    public long mId;
    private Song mSong;

    public int mPlayCount;
    public int mSkipCount;

    public void setPlayCountIfOver(int playCount) {
        if (playCount <= mPlayCount) {
            return;
        }
        mPlayCount = playCount;
    }

    public void setSkipCountIfOver(int skipCount) {
        if (skipCount <= mSkipCount) {
            return;
        }
        mSkipCount = skipCount;
    }

    public void incrementPlayCount(int playCount) {
        mPlayCount += playCount;
    }

    public void incrementSkipCount(int skipCount) {
        mSkipCount += skipCount;
    }

    public void parseSongQueue(SongHistory songHistory) {
        switch (songHistory.getRecordType()) {
            case PLAY:
                setValues(songHistory.getSong(), 1, 0);
                break;
            case SKIP:
                setValues(songHistory.getSong(), 0, 1);
                break;
            default:
                setValues(songHistory.getSong(), 0, 0);
                break;
        }
    }

    public void setValues(Song song, int playCount, int skipCount) {
        setSong(song);
        setPlayCount(playCount);
        setSkipCount(skipCount);
    }
}

File: app/src/main/java/com/sjn/stamp/model/TotalSongHistory.kt
package com.sjn.stamp.model

import com.sjn.stamp.model.constant.RecordType
import io.realm.RealmObject
import io.realm.RealmResults
import io.realm.annotations.LinkingObjects
import io.realm.annotations.PrimaryKey

open class TotalSongHistory(
        @PrimaryKey var id: Long = 0,
        @LinkingObjects("totalSongHistory") val song: RealmResults<Song>? = null,
        var playCount: Int = 0,
        private var skipCount: Int = 0,
        private var completeCount: Int = 0
) : RealmObject() {

    fun merge(src: TotalSongHistory): TotalSongHistory {
        incrementPlayCount(src.playCount)
        incrementSkipCount(src.skipCount)
        incrementCompleteCount(src.completeCount)
        return this
    }

    fun increment(recordType: RecordType): TotalSongHistory {
        when (recordType) {
            RecordType.PLAY -> incrementPlayCount(1)
            RecordType.SKIP -> incrementSkipCount(1)
            RecordType.COMPLETE -> incrementCompleteCount(1)
            else -> {
            }
        }
        return this
    }

    private fun incrementPlayCount(playCount: Int) {
        this.playCount += playCount
    }

    private fun incrementSkipCount(skipCount: Int) {
        this.skipCount += skipCount
    }

    private fun incrementCompleteCount(completeCount: Int) {
        this.completeCount += completeCount
    }
}
